File: my-folder/0239-sliding-window-maximum/solution.py
class Solution(object):
    #que is used to store indices of the array
    def maxSlidingWindow(self, nums, k):
        l,r=0,0
        que=deque()
        res=[]
        while(r<len(nums)):
            while(que and nums[que[-1]]<nums[r]):
                que.pop()
            que.append(r)
            if l>que[0]:
                que.popleft() 
                #when l passes the indices that's stored on top of the queue
            if r+1>=k:
                res.append(nums[que[0]])
                l+=1
            r+=1
        return res


        

File: my-folder/0239-sliding-window-maximum/solution.java
import java.util.*;

class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> dq = new ArrayDeque<>();
        int[] res = new int[nums.length - k + 1];
        int idx = 0;

        for (int r = 0; r < nums.length; r++) {
            // if there are any elements that are not needed within the window range- remove it
            if(!dq.isEmpty() && dq.peekFirst() < r - k + 1){
                dq.pollFirst();
            }

            while(!dq.isEmpty() && nums[dq.peekLast()] < nums[r]){
                dq.pollLast();
            }

            dq.offerLast(r);

            if(r >= k - 1){
                res[idx++] = nums[dq.peekFirst()];
            }
        }

        return res;
    }
}
